### PR TITLE
feat: add `airflow.kubernetesPodTemplate.shareProcessNamespace` value

### PR DESCRIPTION
**Pull request by mthoretton**
## What issues does your PR fix?

I was looking for a nice way to deal with secrets in airflow. As we already use Hashicorp Vault everywhere and are very happy with it I was looking for a good way to make airflow work with it.

There is a [Vault secret backend](https://airflow.apache.org/docs/apache-airflow-providers-hashicorp/stable/secrets-backends/hashicorp-vault.html#hashicorp-vault-secrets) in Airflow but it's very basic. The feature I mostly want to use is to get dynamic credentials for accessing databases, with credentials that will live the time of the pod (dag). For the rest, airflow vars and connections are fine.

I tried to use [Vault CSI](https://github.com/hashicorp/vault-csi-provider) but it's still early stage and also does not fulfill what I want to do. In the end, I just did what I do in other projects: using pod annotations to let vault inject a sidecar in the pod that will inject credentials and revokes them when the pod dies.

Unfortunately, this vault sidecar is meant to always run (running along with a web server or so) and does not work with k8s jobs would end. In order to make it work, I have to add yet another sidecar that will detect when the dag is done then send a signal to kill the vault container. Otherwise, the Pod containing the dag will move to an "unready" state forever as the vault container never dies. [relevant issue in vault project](https://github.com/hashicorp/vault/issues/11089)

I know this is quite a niche PR and quite hacky so you'll decide if it's worth merging. Maybe this PR could still help some people trying to do the same. I guess a proper way could be to change the airflow controller to kill the pod when it detects the Dag is done 🤷 

__Example using new `airflow.kubernetesPodTemplate.shareProcessNamespace` value:__

```yaml
airflow:
  [...]
  kubernetesPodTemplate:
    [...]
      # new value from this PR
      shareProcessNamespace: true

      # sidecar Vault container to inject secrets
      # (uses the shared process namepsace to kill itself when the task finishes)
      extraContainers:
        - name: vault-killer
          image: ubuntu
          command:
            - /bin/bash
            - '-c'
            - 'sleep 5 && while [[ $(pidof dumb-init) ]]; do sleep 5; done; kill -SIGTERM $(pidof vault) || echo "Vault is already gone."'
          resources:
            requests:
              cpu: 30m
              memory: 50Mi
            limits:
              cpu: 100m
              memory: 100Mi
```

## What does your PR do?

- Adds a `airflow.kubernetesPodTemplate.shareProcessNamespace` value to allow enabling the [shareProcessNamespace](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) feature on the KubernetesExecutor pod_template.

### For all Pull Requests

- [x] Commits are [signed off](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#sign-your-work)
- [x] Documentation [updated](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#documentation)
- [x] Passes [ct linting](https://github.com/airflow-helm/charts/tree/main/charts/airflow/CONTRIBUTING.md#linting)

### CODE REVIEW

File: charts/airflow/CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,11 @@ This project follows [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
 
 TBD
 
+## [8.6.0] - 2021-08-25
```
Comment (thesuperzapper) [inline charts/airflow/CHANGELOG.md:10]: Please don't update the changelog, we only do this in the "release commit"

File: charts/airflow/files/pod_template.kubernetes-helm-yaml [hunk 2]
```diff
@@ -66,6 +67,10 @@ spec:
       ports: []
       command: []
       args: []
+      {{- if .Values.airflow.kubernetesPodTemplate.securityContext }}
```
Comment (thesuperzapper) [inline charts/airflow/files/pod_template.kubernetes-helm-yaml:70]: This is not valid, the `airflow.kubernetesPodTemplate.securityContext` value sets the [PodSecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#podsecuritycontext-v1-core), not the [SecurityContext](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.20/#securitycontext-v1-core).

File: charts/airflow/values.yaml [hunk 3]
```diff
@@ -328,6 +328,13 @@ airflow:
     ##
     extraVolumes: []
 
+    ## Enable shareProcessNamespace
+    ## https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/
+    shareProcessNamespace: false
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:333]: Please format these lines like the other docstrings.

File: charts/airflow/values.yaml [hunk 4]
```diff
@@ -328,6 +328,13 @@ airflow:
...
+
+    ## Add extra containers
+    extraContainers: {}
```
Comment (thesuperzapper) [inline charts/airflow/values.yaml:336]: Please format these lines like the other docstrings.

File: charts/airflow/files/pod_template.kubernetes-helm-yaml [hunk 5]
```diff
@@ -74,6 +79,9 @@ spec:
       volumeMounts:
         {{- $volumeMounts | indent 8 }}
       {{- end }}
+    {{- if .Values.airflow.kubernetesPodTemplate.extraContainers }}
+      {{- toYaml .Values.airflow.kubernetesPodTemplate.extraContainers | nindent 4 }}
+    {{- end }}
```
Comment (thesuperzapper) [inline charts/airflow/files/pod_template.kubernetes-helm-yaml:84]: We already have a PR that adds this feature (https://github.com/airflow-helm/charts/pull/456), so we might want to break up your PR and just have it add the `airflow.kubernetesPodTemplate.shareProcessNamespace` value.

File: charts/airflow/values.yaml (hunk 4 above)
Comment (thesuperzapper) [reply charts/airflow/values.yaml:336]: Also, we already have a PR that adds this feature: #456

**Review by thesuperzapper — CHANGES_REQUESTED** [review]
@mthoretton thanks for the PR, but I think this should be scoped down to just adding the `airflow.kubernetesPodTemplate.shareProcessNamespace` value, as PR #456 already adds the `airflow.kubernetesPodTemplate.extraContainers` value.

**Comment by tarekabouzeid** [comment]
Hi,

is there any plans for merging this PR soon ?

Best Regards,

File: charts/airflow/values.yaml (hunk 4 above)
Comment (thesuperzapper) [reply charts/airflow/values.yaml:336]: As https://github.com/airflow-helm/charts/pull/456 will add this feature, please remove this line.

File: charts/airflow/files/pod_template.kubernetes-helm-yaml (hunk 5 above)
Comment (thesuperzapper) [reply charts/airflow/files/pod_template.kubernetes-helm-yaml:84]: As https://github.com/airflow-helm/charts/pull/456 will add this feature, please remove these changes.

File: charts/airflow/files/pod_template.kubernetes-helm-yaml (hunk 2 above)
Comment (thesuperzapper) [reply charts/airflow/files/pod_template.kubernetes-helm-yaml:70]: Please remove this change.

File: charts/airflow/CHANGELOG.md (hunk 1 above)
Comment (thesuperzapper) [reply charts/airflow/CHANGELOG.md:10]: Please remove all updates to the changelog (we only do this for commits that trigger releases).

File: charts/airflow/values.yaml (hunk 3 above)
Comment (thesuperzapper) [reply charts/airflow/values.yaml:333]: For example:
```yaml
    ## sets the `shareProcessNamespace` field on the Pod
    ##  - https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/
    ##
    shareProcessNamespace: false
```

**Comment by thesuperzapper** [comment]
@tarekabouzeid for this PR to be merged:
1. @mthoretton need to update it with the requested changes (see review comments)
2. I need to understand why the ability to set [shareProcessNamespace](https://kubernetes.io/docs/tasks/configure-pod-container/share-process-namespace/) is important, and be confident that this is a use-case we want to enable.
     - @tarekabouzeid feel free to explain your usecase

**Comment by tarekabouzeid** [comment]
@thesuperzapper We have a use case where we are using a sidecar container running in same pods as webserver and scheduler, this container is mounting dags folder from S3 bucket. We are having lifecycle hooks for the sidecar to terminate when the airflow process is terminated so the entire pod is terminated as well. We worked around that for now and its working fine, but it would be much better if we are able to see airflow process from the sidecar.

**Comment by thesuperzapper** [comment]
> @thesuperzapper We have a use case where we are using a sidecar container running in same pods as webserver and scheduler, this container is mounting dags folder from S3 bucket. We are having lifecycle hooks for the sidecar to terminate when the airflow process is terminated so the entire pod is terminated as well. We worked around that for now and its working fine, but it would be much better if we are able to see airflow process from the sidecar.

@tarekabouzeid I don't understand in what situation you would NOT want the containers to be restarted if they exit before they are told to by Kubernetes. When would Kubernetes itself not be the thing terminating the Pod (and thus killing all the containers in it)?

Also, an `s3-sync` sidecar is just like our `git-sync` sidecar, and that runs just fine without knowing anything about the state of the airflow container. 

_PS: I actually want to implement a native `s3-sync` sidecar in the chart, see the issue https://github.com/airflow-helm/charts/issues/249 if you want to help out._

**Comment by mthoretton** [comment]
Sorry for the delay, I think I applied all your comments 👌

**Review by thesuperzapper — APPROVED** [review]
@mthoretton I rebased and did some small cleanup, but it should be ready to merge now

**Comment by tarekabouzeid** [comment]
> > @thesuperzapper We have a use case where we are using a sidecar container running in same pods as webserver and scheduler, this container is mounting dags folder from S3 bucket. We are having lifecycle hooks for the sidecar to terminate when the airflow process is terminated so the entire pod is terminated as well. We worked around that for now and its working fine, but it would be much better if we are able to see airflow process from the sidecar.
> 
> @tarekabouzeid I don't understand in what situation you would NOT want the containers to be restarted if they exit before they are told to by Kubernetes. When would Kubernetes itself not be the thing terminating the Pod (and thus killing all the containers in it)?
> 
> Also, an `s3-sync` sidecar is just like our `git-sync` sidecar, and that runs just fine without knowing anything about the state of the airflow container.
> 
> _PS: I actually want to implement a native `s3-sync` sidecar in the chart, see the issue #249 if you want to help out._

Hi @thesuperzapper , thanks for merging this, we used rclone mount as a side car and mount it inside the webserver and scheduler dags directory. we noticed in case of webserver or scheduler failure the rclone container will keep running and pod wont be terminated. So we added a liveness probe for the rclone container to keep checking if the airflow process in the webserver or scheduler is still running , for that we wanted to have shareProcessNamespace. 
Regarding mounting dags from s3 or gcs ,  rclone solution works fine for now, i checked your comment also about using FUSE , i will look into that.

Best Regards,

**Comment by thesuperzapper** [comment]
@tarekabouzeid I really want to implement built-in support for syncing DAG definitions from S3 and GCS buckets (just like we have for git-sync), if you want to help with this, here are the relevant issues:
- https://github.com/airflow-helm/charts/issues/249
- https://github.com/airflow-helm/charts/issues/561

**Pull request merged**